File: 031_Longest Substring Without Repeating Characters.cpp
/*
Longest Substring Without Repeating Characters

Given a string s, find the length of the longest substring without repeating characters.
*/

#include <bits/stdc++.h>
using namespace std;

int lengthOfLongestSubstring(const string& s) {
    vector<int> last(256, -1);
    int start = 0, best = 0;
    for (int i = 0; i < (int)s.size(); ++i) {
        start = max(start, last[(unsigned char)s[i]] + 1);
        best = max(best, i - start + 1);
        last[(unsigned char)s[i]] = i;
    }
    return best;
}

int main() {
    cout << "Longest 1: " << lengthOfLongestSubstring("abcabcbb") << endl; // 3
    cout << "Longest 2: " << lengthOfLongestSubstring("bbbbb") << endl; // 1
    cout << "Longest 3: " << lengthOfLongestSubstring("pwwkew") << endl; // 3
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1) (fixed 256-size array)

File: 031_Longest Substring Without Repeating Characters.py
"""
Longest Substring Without Repeating Characters

Given a string s, find the length of the longest substring without repeating characters.
"""

def length_of_longest_substring(s: str) -> int:
    last = {}
    start = 0
    best = 0
    for i, ch in enumerate(s):
        if ch in last and last[ch] >= start:
            start = last[ch] + 1
        best = max(best, i - start + 1)
        last[ch] = i
    return best


if __name__ == '__main__':
    print('Longest 1:', length_of_longest_substring('abcabcbb'))
    print('Longest 2:', length_of_longest_substring('bbbbb'))
    print('Longest 3:', length_of_longest_substring('pwwkew'))

# Time Complexity: O(n)
# Space Complexity: O(min(n, charset))
